subprojects/base-services-groovy/src/main/java/org/gradle/api/internal/ClosureBackedAction.java

```diff
@@ -62,13 +62,19 @@ public void execute(T delegate) {
             if (configureableAware && delegate instanceof Configurable) {
                 ((Configurable) delegate).configure(closure);
             } else {
-                Closure copy = (Closure) closure.clone();
-                copy.setResolveStrategy(resolveStrategy);
-                copy.setDelegate(delegate);
-                if (copy.getMaximumNumberOfParameters() == 0) {
-                    copy.call();
-                } else {
-                    copy.call(delegate);
+                int previousResolveStrategy = closure.getResolveStrategy();
```

Comment:
There may be a thread-safety issue with this change. We don't know where the closure comes from, so potentially, it can be affected to multiple projects built in parallel:

```
allprojects {
   tasks.withType(Foo) { 
      doFirst { // this may be executed in parallel }
   }
}
```
:x: